go-controller/pkg/ovn/policy.go

```diff
@@ -685,312 +822,288 @@ func podDeleteAllowMulticastPolicy(nbClient libovsdbclient.Client, ns string, po
-	portGroupIngressDenyName, portGroupEgressDenyName string, obj interface{}) error {
-	np.RLock()
-	defer np.RUnlock()
-	if np.deleted {
+	pgKey := np.namespace
+	// this lock guarantees that sharedPortGroup counters and np.localPods will be updated together
+	// with deleting port from port group atomically
+	oc.sharedNetpolPortGroups.LockKey(pgKey)
+	defer oc.sharedNetpolPortGroups.UnlockKey(pgKey)
+	sharedPGs, ok := oc.sharedNetpolPortGroups.Load(pgKey)
+	if !ok {
+		// Port group doesn't exist, nothing to clean up
+		klog.Infof("Skip delete ports from default deny port group: port group doesn't exist")
 		return nil
 	}
 
-	policyPorts, ingressDenyPorts, egressDenyPorts := oc.processLocalPodSelectorSetPods(policy, np, obj)
+	ingressDenyPorts, egressDenyPorts := sharedPGs.updateDefaultDenyPortsCountersDel(np, portNamesToUUIDs)
+	// counters were updated, update back to initial values on error
+	defer func() {
+		if err != nil {
+			sharedPGs.updateDefaultDenyPortsCountersAdd(np, portNamesToUUIDs)
+		}
+	}()
 
-	var errs []error
+	if len(ingressDenyPorts) != 0 || len(egressDenyPorts) != 0 {
+		// db changes required
+		ops, err = libovsdbops.DeletePortsFromPortGroupOps(oc.nbClient, ops, ingressDenyPGName, ingressDenyPorts...)
+		if err != nil {
+			return fmt.Errorf("unable to get del ports from %s port group ops: %v", ingressDenyPGName, err)
+		}
 
-	ops, err := libovsdbops.AddPortsToPortGroupOps(oc.nbClient, nil, portGroupIngressDenyName, ingressDenyPorts...)
-	if err != nil {
-		oc.processLocalPodSelectorDelPods(np, obj)
-		errs = append(errs, err)
+		ops, err = libovsdbops.DeletePortsFromPortGroupOps(oc.nbClient, ops, egressDenyPGName, egressDenyPorts...)
+		if err != nil {
+			return fmt.Errorf("unable to get del ports from %s port group ops: %v", egressDenyPGName, err)
+		}
+		_, err = libovsdbops.TransactAndCheck(oc.nbClient, ops)
+		if err != nil {
+			return fmt.Errorf("unable to transact del ports from default deny port groups: %v", err)
+		}
 	}
+	return nil
+}
 
-	ops, err = libovsdbops.AddPortsToPortGroupOps(oc.nbClient, ops, portGroupEgressDenyName, egressDenyPorts...)
-	if err != nil {
-		oc.processLocalPodSelectorDelPods(np, obj)
-		errs = append(errs, err)
+// handleLocalPodSelectorAddFunc adds a new pod to an existing NetworkPolicy, should be retriable.
+// ignoreErr=true should only be used for initial objects handling, relying on per-object handlers to retry.
+func (oc *Controller) handleLocalPodSelectorAddFunc(np *networkPolicy, ignoreErr bool, objs ...interface{}) error {
+	np.RLock()
+	defer np.RUnlock()
+	if np.deleted {
+		return nil
+	}
+	// get info for new pods that are not listed in np.localPods
+	portNamesToUUIDs, policyPortUUIDs, errPods := oc.getPolicyPortsForPodsAdd(np, objs...)
+	if !ignoreErr && len(errPods) > 0 {
+		pod := errPods[0].(*kapi.Pod)
+		return fmt.Errorf("unable to get port info for pod %s/%s", pod.Namespace, pod.Name)
```

Comment:
wouldn't we want to add the ports that we were able to get and then return an error after that for hte ones we couldn't? Seems like a best effort operation here would be better. At least we network policy to some of the pods and a single failure doesnt stop NP from being enforced on the rest.

Reply:
`ignoreErr` is exactly for that purpose; it is false for pod events, and true for the initial syncFunc.
But syncFunc is updated in the last commit like this https://github.com/ovn-org/ovn-kubernetes/blob/068bd81fcc28809194b7985c516316b0cf4646a7/go-controller/pkg/ovn/policy.go#L1164-L1166

Reply:
but even without ignoreErr being true (the non sync case), wouldnt we want to apply as much network policy as possible and then still report an error so the rest is retried later? This ensures our network policy that can be applied is done asap.

Reply:
the only case where this function is called for more than 1 object is initial sync for handler, so we know for now if `ignoreErr` is false, there will be only 1 object.
But it makes sense to update this function and ignore this assumption.

Reply:
thanks